pytorch_lightning/trainer/training_loop.py

```diff
@@ -676,26 +678,33 @@ def run_training_batch(self, batch, batch_idx, dataloader_idx):
                     model = self.trainer.get_model()
                     model.toggle_optimizer(optimizer, opt_idx)
 
-                if not (accumulation_done or is_final_batch):
+                if should_accumulate:
                     # For gradient accumulation
 
                     # -------------------
                     # calculate loss (train step + train step end)
                     # -------------------
-                    self.training_step_and_backward(split_batch, batch_idx, opt_idx, optimizer, self.trainer.hiddens)
+
+                    # perform dpp sync only when performing optimizer_step
+                    with self.block_ddp_sync_behaviour():
+                        self.training_step_and_backward(split_batch, batch_idx, opt_idx, optimizer, self.trainer.hiddens)
+
```

Comment:
@justusschock @awaelchli @williamFalcon I think the `backward` in the DDP accelerator should use the wrapped DDP model, not the module inside. Otherwise `require_backward_grad_sync=False` on the wrapped DDP model is being ignored and we miss the grad accumulation benefits

https://github.com/PyTorchLightning/pytorch-lightning/blob/41de4538aa0c187793709a93875e67666c2ddde8/pytorch_lightning/accelerators/accelerator.py#L89-L101

https://github.com/PyTorchLightning/pytorch-lightning/blob/41de4538aa0c187793709a93875e67666c2ddde8/pytorch_lightning/trainer/connectors/model_connector.py#L54-L57

Reply:
https://github.com/PyTorchLightning/pytorch-lightning/pull/4415

Reply:
This was resolved after realising LightningDistributedDataParallel/LightningDataParallel do not override the BW pass